### PR TITLE
kubelet/rkt: Add pullImage(). Fix cat-manifest and others.

### PR DESCRIPTION
**Pull request by yifan-gu**
Add pullImage().
Fix a few bugs in getting pod status.
Fix GetContainerLogs().
Merging envs instead of overriding.
Now being able to run examples/cassandra, examples/secrets.

### CODE REVIEW

**Comment by thockin** [comment]
Dawn: need a ruling on this wrt v1 - my inclination is to shelve it until post 1.0

**Comment by yifan-gu** [comment]
@thockin I hope such fixes can get in as they are not on the common code path(so won't affect other things, safer to merge). And I received several questions recently asking about the status of k8s/rkt, so I hope we can have run some more serious examples at 1.0.

But I absolutely understand it will distract us a little bit and cost a few review bandwidth :P

**Comment by yifan-gu** [comment]
Or maybe @bakins @philips @jonboulle can help review this. And we can merge once it's green?

**Comment by thockin** [comment]
I'm on-call today, so I am making no value judgement on this PR - just
routing traffic

**Comment by erictune** [comment]
Thank you very much for the pull request - we sincerely appreciate the work that goes into sending us a change.

We are currently in a feature-freeze and general code slush as we work toward our 1.0 release. Our initial triage indicates that this is a non-critical change at this time, so we'd like to leave it open and revisit it after 1.0. We expect the freeze to wrap up in early to mid July.

If you feel that this triage is incorrect, please respond and let us know why. We're currently accepting changes that meet any of the following criteria:

Changes directly related to v1.0 issues: https://github.com/GoogleCloudPlatform/kubernetes/issues?q=is%3Aopen+is%3Aissue+milestone%3Av1.0
Changes addressing P0 bugs: https://github.com/GoogleCloudPlatform/kubernetes/issues?q=is%3Aopen+is%3Aissue+label%3Apriority%2FP0
Changes to supported platforms to bring them up to par with the primary development platforms
Docs, user experience, and test related fixes
Thanks again for your contribution! The community is really what has made this project so exciting.

**Comment by jonboulle** [comment]
@yifan-gu fwiw this lgtm

**Comment by yifan-gu** [comment]
I am working on this. I got some examples running with the newer rkt version (>=0.7). But guestbooks examples fails due to an [docker2aci issue](https://github.com/appc/docker2aci/issues/73). I am also going to add per container log features.

**Comment by yifan-gu** [comment]
@dchen1107 I think we can review this and get a merge, otherwise I am afaid the PR will get too big (maybe it's already big...)

**Comment by yifan-gu** [comment]
Basically this PR:
- Update the rkt cli to support rkt 0.7.0. 
- Fix a few bugs 
- Add support for getting per container log
- Add support for image pulling
- Add support for event recording.

File: pkg/kubelet/rkt/pod_info.go [hunk 1]
```diff
@@ -110,14 +110,24 @@ func getIPFromNetworkInfo(networkInfo string) string {
 	return ""
 }
 
+// findExitCode iterates on the exit code map and returns the exit code
+// corresponding to the appName once found. Note that appName is the
```
Comment (jonboulle) [inline pkg/kubelet/rkt/pod_info.go]: unfinished sentence

File: pkg/kubelet/rkt/rkt.go [hunk 2]
```diff
@@ -132,10 +135,13 @@ func New(config *Config,
 		return nil, fmt.Errorf("cannot connect to dbus: %v", err)
 	}
 
-	// Test if rkt binary is in $PATH.
-	rktBinAbsPath, err := exec.LookPath(rktBinName)
+	rktBin := os.Getenv("RKT")
+	if rktBin == "" {
+		return nil, fmt.Errorf("empty 'RKT' environment variable", err)
```
Comment (jonboulle) [inline pkg/kubelet/rkt/rkt.go]: superfluous "err"

File: pkg/kubelet/rkt/rkt.go [hunk 3]
```diff
@@ -512,98 +567,140 @@ func (r *runtime) preparePod(pod *api.Pod) (string, bool, error) {
...
+	var runPrepared string
+	if pod.Spec.HostNetwork {
+		runPrepared = fmt.Sprintf("%s run-prepared --mds-register=false %s", r.rktBinAbsPath, uuid)
+	} else {
+		runPrepared = fmt.Sprintf("%s run-prepared --mds-register=false --private-net %s", r.rktBinAbsPath, uuid)
+	}
 	units := []*unit.UnitOption{
 		newUnitOption(unitKubernetesSection, unitRktID, uuid),
 		newUnitOption(unitKubernetesSection, unitPodName, string(b)),
+		// This makes the service show up for systemd list-unit even it exited successfully.
```
Comment (jonboulle) [inline pkg/kubelet/rkt/rkt.go]: even it -> even if it

File: pkg/kubelet/rkt/rkt.go [hunk 4]
```diff
@@ -512,98 +567,140 @@ func (r *runtime) preparePod(pod *api.Pod) (string, bool, error) {
...
+			continue
+		}
+
+		switch reason {
+		case "created":
+			r.recorder.Eventf(ref, "created", "Created with %q, unit file %q", id, unit)
+		case "started":
+			r.recorder.Eventf(ref, "started", "Started with %q, unit file %q", id, unit)
+		case "failed":
+			r.recorder.Eventf(ref, "failed", " Failed to start the pod %q with error %v", id, failure)
```
Comment (jonboulle) [inline pkg/kubelet/rkt/rkt.go]: " Failed -> "Failed

**Comment by yifan-gu** [comment]
Will need a rebase once #9452 merged.

**Comment by dchen1107** [comment]
Per offline discussion, @yifan-gu is going to split this to small prs for easy reviewing. Thanks!

**Comment by lleszczu** [comment]
I was trying these patches, and due to rkt bug in 0.7.0 (https://github.com/coreos/rkt/pull/1128) it does not work. Additionally it misses option for running custom stage1 aci, so it not possible to run it with latest kvm backend ( https://github.com/coreos/rkt/blob/master/Documentation/running-lkvm-stage1.md ). Can you add option for that ?

**Comment by yifan-gu** [comment]
> Can we use kubelet flag for reading rkt path instead of env variable ?

Sure

> I was trying these patches, and due to rkt bug in 0.7.0 (coreos/rkt#1128) it does not work. Additionally it misses option for running custom stage1 aci, so it not possible to run it with latest kvm backend ( https://github.com/coreos/rkt/blob/master/Documentation/running-lkvm-stage1.md ). Can you add option for that ?

Good catch, I will add this :)

**Comment by yifan-gu** [comment]
As a reminder, add some logs to show what systemd service files we are creating.

**Comment by lleszczu** [comment]
@yifan-gu When do you expect to have patches that fix running k8s with rkt 0.8.0 ? I am mostly interested in running rkt + kvm. If you need some help with it I would be happy to help.

**Comment by yifan-gu** [comment]
@lleszczu Is https://github.com/coreos/rkt/issues/1126 the issue you met? We need to fix that in rkt.
For k8s PRS, hopefully the splits of this PR (https://github.com/kubernetes/kubernetes/pull/12832 https://github.com/kubernetes/kubernetes/pull/12872) will be merged by the end of next week.

**Comment by lleszczu** [comment]
@yifan-gu  Yes, looks like it.
as for PRs, awesome news :)

**Comment by yifan-gu** [comment]
closed as all splitted prs are merged.

**Pull request closed without merging**